pkg/controller/v1alpha2/applicationconfiguration/applicationconfiguration.go

```diff
@@ -544,3 +560,40 @@ func eligible(namespace string, ws []v1alpha2.WorkloadStatus, w []Workload) []un
 
 	return eligible
 }
+
+// ApplyOnceOnlyError indicates the resource being applied has no spec change
+// comparing to the existing one.
+type ApplyOnceOnlyError struct{}
+
+func (e *ApplyOnceOnlyError) Error() string {
+	return "skip applying since no change occurs on this resource"
+}
+
+func applyOnceOnly() resource.ApplyOption {
+	return func(ctx context.Context, current, desired runtime.Object) error {
+		c, _ := current.(metav1.Object)
+		d, _ := desired.(metav1.Object)
+		if c == nil || d == nil {
+			return errors.Errorf("invalid object being applied: %q ",
+				desired.GetObjectKind().GroupVersionKind())
+		}
+		cLabels, dLabels := c.GetLabels(), d.GetLabels()
```

Comment:
what will happen if one of them is nil? For example, it's not update, it's create the first time

Reply:
For Create operation, it will not come into this logic.